### PR TITLE
Feature/joindata

### PR DESCRIPTION
**Pull request by tmelliott**

### CODE REVIEW

**Comment by tmelliott** [comment]
- attach attribute of `join_cols` to data (`col_names` vector)
```r
joindata <- function(...) {
    ...
    # BUT also handle the case when by is not specified (i.e., automatic column matching)
    # try using `capture.output()`
    if (auto matching) {
        vars <- capture.output(res <- interpolate(exp, ...))
        import_join_col <- string_manipulation_of(vars)
        origin_join_col <- import_join_col
    } else {
        res <- interpolate(exp, ...)
    }
    attr(res, "join_cols") <- structure(import_join_col, .Names = origin_join_col)
}

attr(data, 'join_cols')
#   x   a 
# "y" "b" 
```

- remember to update the tests

**Comment by yiwen16** [comment]
Hi I have tried with the below code but it comes up with the error "Error in attr(res, "join_cols") <- structure(import_col, .Names = origin_col) : 
  object 'res' not found"

if (origin_join_col == "" & import_join_col == "") {
    vars = testthat::capture_message(res <- interpolate(exp))
    vars = stringi::stri_extract_all_regex(vars, '"\\S+"')
    import_col = list()
    for (i in 1:length(vars[[1]])) {
      import_col = append(import_col, noquote(vars[[1]][i]))
    }
    origin_col = import_col
  } else {
    res <- interpolate(exp)
  }
  attr(res, "join_cols") <- structure(import_col, .Names = origin_col)
}

I have tried to put the adding attribute line into the if else codes but then only the attributes (the join_cols) are returned

**Comment by tmelliott** [comment]
This works for me
```r
x <- tibble(a = 1:10, b = LETTERS[1:10])
y <- tibble(a = 5:10, c = sample(LETTERS, 6))
vars <- capture.output(res <- inner_join(x, y), type = "message")

res
# # A tibble: 6 x 3
#       a b     c
#   <int> <chr> <chr>
# 1     5 E     Q
# 2     6 F     F
# 3     7 G     U
# 4     8 H     K
# 5     9 I     R
# 6    10 J     Z

vars
# [1] "Joining, by = \"a\""
```

**Comment by yiwen16** [comment]
I am not sure but I think when it comes to the code writing function it is returning the last thing that is being evaluated using the code below, so if I let aa = iNZightTools::joindata(a, b, "", "", "inner_join", "left", "right") then aa is returning me the value of vars not res?
  if ((origin_join_col == "") && (import_join_col == "")) {
    vars = capture.output(res <- interpolate(exp), type = "message")
  } else {
    res <- interpolate(exp)
  }

**Comment by yiwen16** [comment]
if I put in the additional line of code "res <- interpolate(exp)" under the if statement then when I let aa = iNZightTools::joindata(a, b, "", "", "inner_join", "left", "right"), then it is returning me the value of res instead of vars
(a and b are dataframe I created)
 if ((origin_join_col == "") && (import_join_col == "")) {
    vars = capture.output(res <- interpolate(exp), type = "message")
    res <- interpolate(exp)
  } else {
    res <- interpolate(exp)
  }

**Comment by tmelliott** [comment]
This is working for me, I don't see how it should be any different to the final function:

```r
joindata <- function(origin_join_col = "", import_join_col = "") {

    x <- tibble(a = 1:10, b = LETTERS[1:10])
    y <- tibble(a = 5:10, b = LETTERS[5:10], c = sample(LETTERS, 6))

    exp <- ~inner_join(x, y)

    if (origin_join_col == "" & import_join_col == "") {
        vars <- capture.output(res <- interpolate(exp), type = "message")
        # why do fancy string manipulation when we can cheat ...
        origin_join_col <- eval(parse(text = gsub(".+ = ", "", vars)))
        import_join_col <- origin_join_col
    } else {
        res <- suppressMessages(interpolate(exp))
    }

    attr(res, "join_cols") <- structure(import_join_col, .Names = origin_join_col)
    res
}

dat <- joindata()
attr(dat, "join_cols")
#   a   b 
# "a" "b" 


dat <- joindata("a", "a")
attr(dat, "join_cols")
#   a 
# "a" 
```

**Comment by yiwen16** [comment]
Hi thank you it finally working I am so happy!! I think I forgot the second last line of code "res" so the function is not returning res itself

File: tests/testthat/test_joins.R [hunk 1]
```diff
@@ -0,0 +1,47 @@
+context("Test data joins")
+
+stripattr <- function(x, attrs = c('code', 'join_cols')) {
+  for (attr in attrs) attributes(x)[[attr]] <- NULL
+  x
+}
+
+d1 = readr::read_csv("~/iNZight/scripts/join.csv")
```
Comment (tmelliott) [inline tests/testthat/test_joins.R]: Tests run in the `./tests/testthat` directory, so these only need to be
```suggestion
d1 <- readr::read_csv("join.csv")
```

File: tests/testthat/test_joins.R [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+d2 = readr::read_csv("~/iNZight/scripts/join2.csv")
```
Comment (tmelliott) [inline tests/testthat/test_joins.R]:
```suggestion
d2 <- readr::read_csv("join2.csv") 
```

File: tests/testthat/test_joins.R [hunk 3]
```diff
@@ -0,0 +1,47 @@
+context("Test data joins")
+
+stripattr <- function(x, attrs = c('code', 'join_cols')) {
+  for (attr in attrs) attributes(x)[[attr]] <- NULL
+  x
+}
+
+d1 = readr::read_csv("~/iNZight/scripts/join.csv")
+d2 = readr::read_csv("~/iNZight/scripts/join2.csv") 
+d3 = readr::read_csv("~/iNZight/scripts/join3.csv")
```
Comment (tmelliott) [inline tests/testthat/test_joins.R]:
```suggestion
d3 <- readr::read_csv("join3.csv") 
```

File: tests/testthat/test_joins.R [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+d1 = readr::read_csv("~/iNZight/scripts/join.csv")
+d2 = readr::read_csv("~/iNZight/scripts/join2.csv") 
+d3 = readr::read_csv("~/iNZight/scripts/join3.csv") 
+
+test_that("Auto detection works", {
+  expect_equal(stripattr(joindata(d1, d2, "", "", "inner_join", "x", "y")), inner_join(d1, d2))
+})
+
+test_that("Inner join works", {
+  expect_equal(stripattr(joindata(d1, d3, list("x1", "x2"), list("x1", "x6"), "inner_join", "x", "y")), inner_join(d1, d3, by = c("x1" = "x1", "x2" = "x6")))
```
Comment (tmelliott) [inline tests/testthat/test_joins.R]:
```suggestion
  expect_equal(
      stripattr(joindata(d1, d3, list("x1", "x2"), list("x1", "x6"), "inner_join", "x", "y")), 
      inner_join(d1, d3, by = c("x1" = "x1", "x2" = "x6"))
  )
```

**Pull request merged**